File: src/main/java/com/stepasha/javaorders/models/Customers.java
package com.stepasha.javaorders.models;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "customers")
public class Customers {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long custcode;
    @Column(nullable = false)
    private String custname;
    private String custcity;
    private String workingarea;
    private String custcountry;
    private String grade;
    @Transient
    public boolean hasvalueforopeningatm = false;
    private double openingamt;
    @Transient
    public boolean hasvalueforreceiveatm = false;
    private double receiveamt;
    @Transient
    public boolean hasvalueforpaymentatm = false;
    private double paymentamt;
    @Transient
    public boolean hasvalueforoutstandingatm =false;
    private double outstandingamt;
    private String phone;

    //many to one
    //many costumers belong to one agent
    @ManyToOne
    @JoinColumn(name = "agentcode", nullable = false)
    @JsonIgnoreProperties("customers")
    private Agents agent;

    //one customer could have many orders
    @OneToMany(mappedBy = "customer", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnoreProperties("customer")
    private List<Orders> orders = new ArrayList<>();

    public Customers() {
    }

    public Customers(String custname,
                    String custcity,
                    String workingarea,
                    String custcountry,
                    String grade,
                    double openingamt,
                    double receiveamt,
                    double paymentamt,
                    double outstandingamt,
                    String phone,
                    Agents agent) {
        this.custname = custname;
        this.custcity = custcity;
        this.workingarea = workingarea;
        this.custcountry = custcountry;
        this.grade = grade;
        this.openingamt = openingamt;
        this.receiveamt = receiveamt;
        this.paymentamt = paymentamt;
        this.outstandingamt = outstandingamt;
        this.phone = phone;
        this.agent = agent;
    }

    public long getCustcode() {
        return custcode;
    }

    public void setCustcode(long custcode) {
        this.custcode = custcode;
    }

    public String getCustname() {
        return custname;
    }

    public void setCustname(String custname) {
        this.custname = custname;
    }

    public String getCustcity() {
        return custcity;
    }

    public void setCustcity(String custcity) {
        this.custcity = custcity;
    }

    public String getWorkingarea() {
        return workingarea;
    }

    public void setWorkingarea(String workingarea) {
        this.workingarea = workingarea;
    }

    public String getCustcountry() {
        return custcountry;
    }

    public void setCustcountry(String custcountry) {
        this.custcountry = custcountry;
    }

    public String getGrade() {
        return grade;
    }

    public void setGrade(String grade) {
        this.grade = grade;
    }

    public double getOpeningamt() {
        return openingamt;
    }

    public void setOpeningamt(double openingamt) {

        //TODO 7 for anything that cant be null
        hasvalueforopeningatm = true;
        this.openingamt = openingamt;
    }

    public double getReceiveamt() {
        return receiveamt;
    }

    public void setReceiveamt(double receiveamt) {

        //TODO 7 for anything that cant be null
        hasvalueforreceiveatm = true;
        this.receiveamt = receiveamt;
    }

    public double getPaymentamt() {
        return paymentamt;
    }

    public void setPaymentamt(double paymentamt) {

        //TODO 7 for anything that cant be null
        hasvalueforpaymentatm = true;
        this.paymentamt = paymentamt;
    }

    public double getOutstandingamt() {
        return outstandingamt;
    }

    public void setOutstandingamt(double outstandingamt) {

        //TODO 7 for anything that cant be null
        hasvalueforoutstandingatm = true;
        this.outstandingamt = outstandingamt;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public Agents getAgent() {
        return agent;
    }

    public void setAgent(Agents agent) {
        this.agent = agent;
    }



    public List<Orders> getOrders() {
        return orders;
    }

    public void setOrders(List<Orders> orders) {
        this.orders = orders;
    }


}

File: src/main/kotlin/com/wkdrabbit/crudycustomers/model/Customers.kt
package com.wkdrabbit.crudycustomers.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import javax.persistence.*

class Customers{

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    val customerId : Long = 0

    @Column(unique = true, nullable = false)
    val name : String = ""


    //TODO: Fix this to match data
    var customerName = ""
    var customerCity = ""
    var workingArea = ""
    var customerCountry = ""
    var grade = ""
    var openingAmount = 0.0
    var receiveAmount = 0.0
    var paymentAmount = 0.0
    var outstandingAmount = 0.0
    var phone = ""

    @ManyToOne
    @JoinColumn(name = "agentId", nullable = false)
    var agent : Agents = Agents()


    @OneToMany(mappedBy = "agent",
            cascade = [CascadeType.ALL],
            orphanRemoval = true)
    @JsonIgnoreProperties("customer")
    val orders = mutableListOf<Orders>()

    constructor(customerName: String, customerCity: String, workingArea: String, customerCountry: String, grade: String, openingAmount: Double, receiveAmount: Double, paymentAmount : Double, outstandingAmount: Double, phone: String, agent: Agents) {
        this.customerName = customerName
        this.customerCity = customerCity
        this.workingArea = workingArea
        this.customerCountry = customerCountry
        this.grade = grade
        this.openingAmount = openingAmount
        this.receiveAmount = receiveAmount
        this.paymentAmount = paymentAmount
        this.outstandingAmount = outstandingAmount
        this.phone = phone
        this.agent = agent
    }

    constructor()


}